internal/servers/worker: Controller failure connection cleanup

vancluever opened the pull request:
This commit adds some rudimentary functionality for cleaning up
connections on controller failure, or more specifically, a failure to
communicate with the controllers.

In this scenario, when a status cannot be sent for any reason for longer
than 30 seconds, the worker will close all connections that are normally
not cleaned up (so for sessions that currently are *not* in the progress
of being canceled or terminated). Status requests now time out after 10
seconds if they are not made successfully.

Session status is not touched during this process, but natural cleanup
or connectionless sessions will still occur. This should be okay, as the
session should be added on re-connect if need be.

A few other changes are made to cancelConnections: first, error logging
has been moved to the function so that we can handle errors requesting
cancellation directly within the function; this entails simply just
locally marking these connections as closed with a warning. Second,
there was a minor logging bug where closed connections were not properly
being noted in the last trace log, this has been fixed.

vancluever commented:
Closing #1257 in favor of this one so that I can start merging other controller-related changes into this one to take advantage of some of the acceptance testing bits, and we can make further enhancements against progress here.

jefferai commented on internal/servers/worker/status.go:
```diff
@@ -115,6 +159,33 @@ func (w *Worker) sendWorkerStatus(cancelCtx context.Context) {
+		// scenario, as there will be no way we can really tell if these
+		// connections should continue to exist.
+
+		lastStatus := w.LastStatusSuccess()
```
I think it's important to capture something more for when lastStatus is nil. It seems _unlikely_ that we'd never have sent status but have connections, but I think what we could do is on worker startup time capture the current time. If lastStatus is nil, use that instead as the comparison for the grace period.

vancluever replied:
Fixed!

jefferai commented on internal/servers/worker/status.go:
```diff
@@ -115,6 +159,33 @@ func (w *Worker) sendWorkerStatus(cancelCtx context.Context) {
+
+		lastStatus := w.LastStatusSuccess()
+		gracePeriod := w.statusGracePeriod()
+		if lastStatus != nil && lastStatus.StatusTime.Before(time.Now().Add(gracePeriod*-1)) {
```
Pedantry: the Before/Add stuff could probably be simplified a bit by using `time.Sub`

vancluever replied:
Apparently I thought this didn't exist, because I swear I looked for it 🤦‍♂️ Will fix.

vancluever replied:
Fixed!

jefferai commented on go.mod:
```diff
@@ -55,6 +55,7 @@ require (
 	github.com/posener/complete v1.2.3
 	github.com/spf13/cobra v1.1.1 // indirect
 	github.com/stretchr/testify v1.7.0
+	github.com/vancluever/dawdle v0.3.0
```
This does need to move into the HC org prior to merging.

vancluever replied:
Done!

jefferai commented on internal/servers/worker/worker.go:
```diff
@@ -28,6 +28,7 @@ type Worker struct {
 	started     *ua.Bool
 
 	controllerStatusConn *atomic.Value
+	workerStartTime      *atomic.Value
```
This doesn't need to be an atomic value since it will never change -- it will be set on startup and only read thereafter. It can be a plain `time.Time`.

vancluever replied:
Fixed!

vancluever commented on internal/servers/worker/status.go:
```diff
@@ -154,46 +225,98 @@ func (w *Worker) sendWorkerStatus(cancelCtx context.Context) {
+
+type cleanupConnectionsCondition func(si *sessionInfo) bool
+
+func cleanupConnectionsConditionAnyStatus(statuses ...services.SESSIONSTATUS) cleanupConnectionsCondition {
```
Fixed!

vancluever commented:
Addressed all current review comments, and as discussed, removed (for now) the unit testing done in `internal/servers/worker`. These tests may come back at a later time if they're deemed to be particularly useful outside of what we have in E2E tests and `TestWorker` already.

vancluever commented:
@jefferai this should be good to go now.

jefferai commented on internal/tests/cluster/session_cleanup_test.go:
```diff
@@ -0,0 +1,509 @@
+package cluster
```
Overall (maybe nit) on the file: Convention for us is not to pass require objects around but rather testing.T objects. Partly it's due to Go convention, partly because the only option with require is to fail immediately -- it precludes you from deciding you want to assert without failing, and partly it's because you can't call `t.Log` and friends since you don't have access to the original testing.T.

jefferai reviewed (approved):
Approved pending addressing my one comment
